TypeCobol.LanguageServer/Workspace.cs

```diff
@@ -951,6 +952,189 @@ private void FinalCompilationStepCompleted(object cUnit, ProgramClassEvent progr
+        /// </summary>
+        /// <param name="compilationUnit">Compilation unit resulting from parsing the Program/Copy</param>
+        /// <param name="separator">Separator to be used in the rows</param>
+        /// <returns></returns>
+        public string[] GetDataLayoutAsCSV(CompilationUnit compilationUnit, string separator)
+        {
+            var rows = new List<string>();
+            foreach (var dataLayoutNode in CollectDataLayoutNodes(compilationUnit))
+            {
+                var row = CreateRow(dataLayoutNode, separator);
+                if (row != null)
+                {
+                    rows.Add(row);
+                }
+            }
+
+            return rows.ToArray();
+
+            static string CreateRow(Tuple<int, DataDefinition> dataLayoutNode, string separator)
+            {
+                var nodeLevel = dataLayoutNode.Item1;
+                var dataDefinition = dataLayoutNode.Item2;
+
+                var row = new StringBuilder();
+
+                //TODO manage slack bytes (property is dataDefinition.SlackBytes)
+
+                // Line number (starting at 1)
+                AppendToRow(dataDefinition.CodeElement.GetLineInMainSource() + 1);
+
+                /// Level number
+                AppendToRow(dataDefinition.CodeElement.LevelNumber);
+
+                // Name (preceded by an indent depending on the node level)
+                row.Append(SPACE, nodeLevel * INDENT_SIZE);
+                AppendToRow(GetName(dataDefinition));
+
+                // Declaration (Picture, Usage, REDEFINES, OCCURS, ...)
+                AppendDeclarationToRow(dataDefinition);
+
+                // Start/End/Length
+                var start = dataDefinition.StartPosition;
+                var length = dataDefinition.PhysicalLength;
+                AppendToRow(start);
+                AppendToRow(GetEnd(start, length));
+                AppendToRow(length);
+
+                return row.ToString();
+
+                void AppendToRow(object value)
+                {
+                    row.Append(value).Append(separator);
+                }
+
+                static string GetName(DataDefinition dataDefinition)
+                {
+                    var name = dataDefinition.Name ?? FILLER;
+                    if (dataDefinition.IsUnderTable)
+                    {
+                        name += OCCURS_SUFFIX;
+                    }
+                    return name;
+                }
+
+                void AppendDeclarationToRow(DataDefinition dataDefinition)
+                {
+                    var initialRowLength = row.Length;
+
+                    if (dataDefinition is DataRedefines dataRedefines)
+                    {
+                        row.Append(string.Format(REDEFINES, dataRedefines.CodeElement.RedefinesDataName.Name));
+                    }
+                    else if (dataDefinition.CodeElement is DataDescriptionEntry dataDescriptionEntry)
```

Comment:
Can the type checks be replaced with CodeElementType comparisons ?

Reply:
Yes, done